ugvc/cnv/normalize_reads_count.R

```diff
@@ -0,0 +1,314 @@
+
+		if (ncol(X)==1){
+			Y <- X
+		} else {
+
+			Y <- matrix(0,nrow=nrow(X),ncol=ncol(X))
+			chrIdx <- which((chr != "chrX") & (chr != "chrY"))
+			Ytmp <- X[chrIdx, ,drop=FALSE]
+
+			if (all(Ytmp==0)){
+				Y[chrIdx, ] <- Ytmp
+			} else {
+				idxSG <- apply(Ytmp,1,function(x) all(x<1))
+				Ytmp[idxSG, ] <- NA
+
+				if ((nrow(Ytmp)-length(which(idxSG))) > 1){
+
+					if (sizeFactor=="mean"){
+						sizeFactors <- colMeans(Ytmp,na.rm=TRUE)
+					} else if (sizeFactor=="median" | normType=="poisson"){
+						sizeFactors <- apply(Ytmp,2,median,na.rm=TRUE)
+					} else if (sizeFactor=="quant"){
+						sizeFactors <- apply(Ytmp,2,quantile,probs=quSizeFactor,na.rm=TRUE)
+					} else if (sizeFactor=="mode"){
+						sizeFactors <- apply(Ytmp,2,function(x) .statmod(x[x!=0],na.rm=TRUE) )
+					}
+
+					if (any(is.na(sizeFactors))){
+						warning(paste("Normalization failed for reference sequence ",l,
+										". Using global sizeFactors!"))
+						sizeFactors <- globalsizeFactors
+					}
+
+					if (any(sizeFactors==0)){
+						stop(paste("Some normalization factors are zero!",
+										"Remove samples or chromosomes for which the average read count is zero,",
+										"e.g. chromosome Y."))
+					}
+
+
+						if (normType=="mean"){
+							correctwiththis <-  mean(sizeFactors,na.rm=TRUE)/sizeFactors
+						} else if (normType=="median"){
+							correctwiththis <-  median(sizeFactors,na.rm=TRUE)/sizeFactors
+						} else if (normType=="quant"){
+							correctwiththis <-  quantile(sizeFactors,probs=qu,na.rm=TRUE)/sizeFactors
+						} else if (normType=="mode"){
+							asm <- .statmod(Ytmp[Ytmp!=0],na.rm=TRUE)
+							correctwiththis <- asm/sizeFactors
+						} else if (normType=="poisson"){
+							correctwiththis <-  mean(sizeFactors,na.rm=TRUE)/sizeFactors
+							#YYtmp <- t(t(Ytmp)*correctwiththis)
+							YYtmp <- Ytmp %*% diag(correctwiththis)
+							v2m <- apply(YYtmp,1,var)/rowMeans(YYtmp)
+							uut <- quantile(v2m,probs=0.95,na.rm=TRUE)
+							dd <- density(v2m,na.rm=TRUE,from=0,
+									to=uut,n=uut*100)
+							#mv2m <- median(v2m,na.rm=TRUE)
+							mv2m <- dd$x[which.max(dd$y)]
+							if (is.finite(mv2m)) {correctwiththis <- correctwiththis*1/mv2m}
+						}
+						#browser()
+
+					} else {
+						warning(paste("Normalization for reference sequence ",l,"not",
+										"applicable, because of low number of segments"))
+						correctwiththis <- rep(1,ncol(X))
+					}
+
+					if (any(!is.finite(correctwiththis))){
+						warning(paste("Normalization for reference sequence ",l,"not",
+										"applicable, because at least one sample has zero",
+										"reads."))
+						correctwiththis <- rep(1,ncol(X))
+					}
+
+					Ytmp <- X %*% diag(correctwiththis)
+					Ytmp[idxSG, ] <- 0
+
+					Y <- Ytmp
+				}
+
+		} # over chr
+
+	#handle Sex chromosomes(chrX,chrY) by Gender ploidy value.
+	#for male we will multiply normalized X and Y coverage with 2.
```

Comment:
Why do we need to multiply Y coverage by 2?

Reply:
in current implementation we need to do this in order to not read chrY as deletion for male samples.

Reply:
I am sorry, I don't understand this. The Y chromosomes are all the same in all males, no? I understand why you would multiply chrX by two in males

Reply:
Since normalization is done Genome-wise, if we will not multiply it  by 2 we will get CN=1 for that chromosome same as for chrX. in our next update, we will try to run chrY separately from the cohort, and then we will not have to multiply by2.